File: ios/Andrews/Model/Pricing.swift
import Foundation

enum Pricing {
    static let taxRate: Decimal = 0.07
    static let deliveryFee = Decimal(sign: .plus, exponent: -2, significand: 299)

    static func itemPrice(_ sel: Selection) -> Decimal {
        let selected = sel.ingredientIds.map(Menu.ingredient)
        let extrasOnly = !selected.isEmpty && selected.allSatisfy { $0.group == .extras }
        let base = extrasOnly ? Decimal.zero : Menu.mealType(sel.mealType).basePrice
        let upcharges = selected.reduce(Decimal.zero) { $0 + $1.price }
        return rounded((base + upcharges) * Decimal(sel.quantity))
    }

    static func bagSubtotal(_ items: [BagItem]) -> Decimal {
        rounded(items.reduce(Decimal.zero) { $0 + itemPrice($1.selection) })
    }

    static func mealCount(_ items: [BagItem]) -> Int {
        items.reduce(0) { $0 + $1.selection.quantity }
    }

    static func planDiscountRate(_ plan: PlanSize, count: Int) -> Decimal {
        guard count >= plan.rawValue else { return 0 }
        switch plan {
        case .five:
            return 0
        case .seven:
            return 0.05
        case .ten:
            return 0.10
        }
    }

    static func promoRate(_ code: String?) -> Decimal {
        code?.trimmingCharacters(in: .whitespacesAndNewlines).uppercased() == "CYCLONE10"
            ? 0.10
            : 0
    }

    /// The delivery fee is added after tax and is not taxed.
    static func orderTotals(
        _ items: [BagItem],
        plan: PlanSize,
        promo: String?,
        deliveryFee: Decimal = 0
    ) -> Totals {
        let subtotalRaw = items.reduce(Decimal.zero) { $0 + itemPrice($1.selection) }
        let discountRate = planDiscountRate(plan, count: mealCount(items)) + promoRate(promo)
        let discountRaw = subtotalRaw * discountRate
        let taxRaw = (subtotalRaw - discountRaw) * taxRate
        return Totals(
            subtotal: rounded(subtotalRaw),
            discount: rounded(discountRaw),
            tax: rounded(taxRaw),
            delivery: rounded(deliveryFee),
            total: rounded(subtotalRaw - discountRaw + taxRaw + deliveryFee)
        )
    }

    /// Names the worst miss only, protein before calories.
    static func goalStatus(
        _ goal: Goal,
        macros: (calories: Int, protein: Int)
    ) -> (onTarget: Bool, message: String) {
        let proteinShort = goal.minimumProtein - macros.protein
        if proteinShort > 0 {
            return (false, "\(proteinShort)g protein short")
        }
        let range = goal.calorieRange
        if macros.calories > range.upperBound {
            return (false, "\(macros.calories - range.upperBound) cal over")
        }
        if macros.calories < range.lowerBound {
            return (false, "\(range.lowerBound - macros.calories) cal under")
        }
        return (true, "On target")
    }

    static func budgetStatus(budget: Decimal, price: Decimal) -> (over: Bool, remaining: Decimal) {
        let remaining = rounded(budget - price)
        return (remaining < 0, remaining)
    }

    static func macros(_ sel: Selection) -> (calories: Int, protein: Int) {
        let sum = sel.ingredientIds.map(Menu.ingredient).reduce(into: (calories: 0, protein: 0)) {
            $0.calories += $1.calories
            $0.protein += $1.protein
        }
        return (sum.calories * sel.quantity, sum.protein * sel.quantity)
    }

    static func money(_ d: Decimal) -> String {
        let formatter = NumberFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.numberStyle = .decimal
        formatter.positivePrefix = "$"
        formatter.negativePrefix = "-$"
        formatter.usesGroupingSeparator = true
        formatter.minimumFractionDigits = 2
        formatter.maximumFractionDigits = 2
        return formatter.string(from: NSDecimalNumber(decimal: rounded(d))) ?? "$0.00"
    }

    private static func rounded(_ value: Decimal) -> Decimal {
        NSDecimalNumber(decimal: value).rounding(
            accordingToBehavior: NSDecimalNumberHandler(
                roundingMode: .plain,
                scale: 2,
                raiseOnExactness: false,
                raiseOnOverflow: true,
                raiseOnUnderflow: true,
                raiseOnDivideByZero: true
            )
        ).decimalValue
    }
}

File: src/lib/pricing.ts
import { ingredient, mealType } from "../data/menu";
import type { BagItem, PlanSize, Selection } from "../data/types";

export type Totals = { subtotal: number; discount: number; tax: number; delivery: number; total: number };
export const TAX_RATE = 0.07;
export const DELIVERY_FEE = 2.99;

const money = (value: number) => Math.round((value + Number.EPSILON) * 100) / 100;

export function itemPrice(sel: Selection): number {
  const selected = sel.ingredientIds.map(ingredient);
  const extrasOnly = selected.length > 0 && selected.every((item) => item.group === "extras");
  const base = extrasOnly ? 0 : mealType(sel.mealType).basePrice;
  return money((base + selected.reduce((sum, item) => sum + item.price, 0)) * sel.quantity);
}

export function bagSubtotal(items: BagItem[]): number {
  return money(items.reduce((sum, item) => sum + itemPrice(item), 0));
}

export function mealCount(items: BagItem[]): number {
  return items.reduce((sum, item) => sum + item.quantity, 0);
}

export function planDiscountRate(plan: PlanSize, count: number): number {
  if (count < plan) return 0;
  if (plan === 5) return 0.15;
  if (plan === 7) return 0.2;
  if (plan === 10) return 0.25;
  return 0;
}

export function promoRate(code: string | undefined): number {
  return code?.trim().toUpperCase() === "CYCLONE10" ? 0.1 : 0;
}

export function orderTotals(items: BagItem[], plan: PlanSize, promo?: string, deliveryFee = 0): Totals {
  const subtotalRaw = items.reduce((sum, item) => sum + itemPrice(item), 0);
  const discountRaw = subtotalRaw * (planDiscountRate(plan, mealCount(items)) + promoRate(promo));
  const taxRaw = (subtotalRaw - discountRaw) * TAX_RATE;
  return {
    subtotal: money(subtotalRaw),
    discount: money(discountRaw),
    tax: money(taxRaw),
    delivery: money(deliveryFee),
    total: money(subtotalRaw - discountRaw + taxRaw + deliveryFee),
  };
}

export function macros(sel: Selection): { calories: number; protein: number } {
  const result = sel.ingredientIds.map(ingredient).reduce(
    (sum, item) => ({ calories: sum.calories + item.calories, protein: sum.protein + item.protein }),
    { calories: 0, protein: 0 },
  );
  return { calories: result.calories * sel.quantity, protein: result.protein * sel.quantity };
}
